File: frontend/src/api/alerts.ts
import { useMemo } from 'react';
import { useMutation, useQuery } from '@tanstack/react-query';
import { toast } from 'sonner';
import { api, getErrorMessage, queryClient } from './client';
import { useEntityInfoMap } from './entityInfo';
import { getEntityTypeLabel } from '@/utils/formatters';
import type {
  AlertBulkUpdate,
  AlertConfig,
  AlertConfigUpdate,
  AlertLog,
  WhatsAppStatus,
  WhatsAppTestResult,
} from '@/types';

interface BackendConfig {
  id: string;
  entity_type: string;
  entity_id: string;
  days_before: number[];
  channels: string[];
  is_active: boolean;
  created_at: string;
}

interface BackendLog {
  id: string;
  entity_type: string;
  entity_id: string;
  channel: string;
  days_before: number | null;
  sent_date: string;
  status: string;
  error_message: string | null;
  sent_at: string;
}

function toConfig(c: BackendConfig | AlertConfig, infoMap: Map<string, { name: string; amount: number; due_date: string }>): AlertConfig {
  // MSW mock mode already returns AlertConfig-shaped objects — pass through.
  if ('enabled' in c) return c;
  const match = infoMap.get(c.entity_id);
  return {
    id: c.id,
    entity_type: c.entity_type as AlertConfig['entity_type'],
    entity_id: c.entity_id,
    entity_name: match?.name ?? getEntityTypeLabel(c.entity_type as AlertConfig['entity_type']),
    due_date: match?.due_date ?? c.created_at.slice(0, 10),
    amount: match?.amount ?? 0,
    enabled: c.is_active,
    channels: c.channels as AlertConfig['channels'],
    days_before: c.days_before as AlertConfig['days_before'],
  };
}

function toLog(l: BackendLog | AlertLog): AlertLog {
  // MSW mock mode already returns AlertLog-shaped objects — pass through.
  if ('message' in l) return l;
  return {
    id: l.id,
    entity_type: l.entity_type as AlertLog['entity_type'],
    entity_id: l.entity_id,
    entity_name: getEntityTypeLabel(l.entity_type as AlertLog['entity_type']),
    channel: l.channel as AlertLog['channel'],
    status: l.status as AlertLog['status'],
    days_before: l.days_before ?? 0,
    message: l.error_message ?? `${l.channel} reminder sent`,
    sent_at: l.sent_at,
  };
}

export const useAlertConfigs = () => {
  // Shared, cached entity-info lookup (see entityInfo.ts) — AlertConfigOut has
  // no entity_name/amount/due_date, and this resolver is deduped across every
  // caller (payments, dashboard, alerts) via a single query key.
  const infoMap = useEntityInfoMap();

  const listQuery = useQuery({
    queryKey: ['alerts', 'configs'],
    queryFn: async () => {
      const { data } = await api.get<{ items: (BackendConfig | AlertConfig)[]; total: number }>(
        '/alerts/configs',
        { params: { limit: 100 } },
      );
      return data.items;
    },
  });

  const data = useMemo(() => {
    if (!listQuery.data) return listQuery.data;
    return listQuery.data.map((c) => toConfig(c, infoMap.data ?? new Map()));
  }, [listQuery.data, infoMap.data]);

  return { ...listQuery, data, isLoading: listQuery.isLoading || infoMap.isLoading };
};

export const useAlertLogs = (entityId?: string) =>
  useQuery({
    queryKey: ['alerts', 'logs', entityId ?? 'all'],
    queryFn: () =>
      api
        .get<{ items: (BackendLog | AlertLog)[]; total: number }>('/alerts/logs', { params: { limit: 100 } })
        .then((r) => r.data.items.map(toLog).filter((l) => !entityId || l.entity_id === entityId)),
  });

export const useUpdateAlertConfig = () =>
  useMutation({
    mutationFn: ({ id, data }: { id: string; data: AlertConfigUpdate }) => {
      const { enabled, ...rest } = data;
      return api
        .patch<BackendConfig>(`/alerts/configs/${id}`, {
          ...rest,
          is_active: enabled,
        })
        .then((r) => r.data);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['alerts', 'configs'] });
    },
    onError: (error) => toast.error(getErrorMessage(error) || 'Could not update alert settings'),
  });

/**
 * Apply one setting to every rule at once. The reminder schedule and channels
 * are a household-wide preference, so the settings page edits them here rather
 * than looping a request per payable.
 */
export const useBulkUpdateAlertConfigs = () =>
  useMutation({
    mutationFn: ({ enabled, ...rest }: AlertBulkUpdate) =>
      api
        .patch<{ updated: number }>('/alerts/configs', {
          ...rest,
          ...(enabled === undefined ? {} : { is_active: enabled }),
        })
        .then((r) => r.data),
    onSuccess: ({ updated }) => {
      queryClient.invalidateQueries({ queryKey: ['alerts', 'configs'] });
      toast.success(updated === 0 ? 'Nothing to update' : `Updated ${updated} reminder rules`);
    },
    onError: (error) => toast.error(getErrorMessage(error) || 'Could not update alert settings'),
  });

/** Whether Twilio can send, and to which (masked) number. */
export const useWhatsAppStatus = () =>
  useQuery({
    queryKey: ['alerts', 'whatsapp'],
    queryFn: () => api.get<WhatsAppStatus>('/alerts/whatsapp').then((r) => r.data),
    staleTime: 5 * 60_000,
  });

/**
 * Send a real WhatsApp message down the same path an alert takes. The backend
 * reports a misconfiguration as `sent: false` with a reason rather than an
 * error status, so the toast can say what to fix.
 */
export const useSendWhatsAppTest = () =>
  useMutation({
    mutationFn: () =>
      api.post<WhatsAppTestResult>('/alerts/whatsapp/test').then((r) => r.data),
    onSuccess: (result) => {
      if (result.sent) toast.success(result.detail);
      else toast.error(result.detail);
    },
    onError: (error) => toast.error(getErrorMessage(error) || 'Could not send the test message'),
  });

File: backend/app/api/v1/alerts.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user, get_db, get_vault_owner_id, require
from app.core.permissions import ALERTS_EDIT, ALERTS_VIEW
from app.models.user import User
from app.schemas.alert import (
    AlertBulkUpdateResult,
    AlertConfigBulkUpdate,
    AlertConfigListResponse,
    AlertConfigOut,
    AlertConfigUpdate,
    AlertLogListResponse,
    WhatsAppStatus,
    WhatsAppTestResult,
)
from app.services import alert_service

router = APIRouter(prefix="/alerts", tags=["alerts"])


@router.get(
    "/configs", response_model=AlertConfigListResponse, dependencies=[Depends(require(ALERTS_VIEW))]
)
async def list_configs(
    entity_type: str | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    vault_id: uuid.UUID = Depends(get_vault_owner_id),
):
    return await alert_service.list_configs(db, vault_id, entity_type, skip, limit)


@router.patch(
    "/configs",
    response_model=AlertBulkUpdateResult,
    dependencies=[Depends(require(ALERTS_EDIT))],
)
async def bulk_update_configs(
    payload: AlertConfigBulkUpdate,
    db: AsyncSession = Depends(get_db),
    vault_id: uuid.UUID = Depends(get_vault_owner_id),
):
    """Apply one setting to every rule at once.

    Declared before `/configs/{config_id}` so the literal path is matched first.
    """
    return await alert_service.bulk_update_configs(db, vault_id, payload)


@router.get(
    "/whatsapp",
    response_model=WhatsAppStatus,
    dependencies=[Depends(require(ALERTS_VIEW))],
)
async def whatsapp_status():
    """Whether Twilio can send, and to which (masked) number. Never returns
    the auth token."""
    return alert_service.whatsapp_status()


@router.post(
    "/whatsapp/test",
    response_model=WhatsAppTestResult,
    dependencies=[Depends(require(ALERTS_EDIT))],
)
async def send_whatsapp_test(current_user: User = Depends(get_current_user)):
    """Send a real WhatsApp message down the same path an alert takes."""
    return await alert_service.send_whatsapp_test(current_user.phone_number)


@router.get(
    "/configs/{config_id}",
    response_model=AlertConfigOut,
    dependencies=[Depends(require(ALERTS_VIEW))],
)
async def get_config(
    config_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    vault_id: uuid.UUID = Depends(get_vault_owner_id),
):
    return await alert_service.get_config(db, vault_id, config_id)


@router.patch(
    "/configs/{config_id}",
    response_model=AlertConfigOut,
    dependencies=[Depends(require(ALERTS_EDIT))],
)
async def update_config(
    config_id: uuid.UUID,
    payload: AlertConfigUpdate,
    db: AsyncSession = Depends(get_db),
    vault_id: uuid.UUID = Depends(get_vault_owner_id),
):
    return await alert_service.update_config(db, vault_id, config_id, payload)


@router.get(
    "/logs", response_model=AlertLogListResponse, dependencies=[Depends(require(ALERTS_VIEW))]
)
async def list_logs(
    entity_type: str | None = Query(None),
    status: str | None = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    vault_id: uuid.UUID = Depends(get_vault_owner_id),
):
    return await alert_service.list_logs(db, vault_id, entity_type, status, skip, limit)
